pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,948 @@
+	// state to store the new BGP server, listen port, local ASN, and router ID.
+	needUpdateBGPServer := c.bgpPolicyState == nil ||
+		c.bgpPolicyState.listenPort != listenPort ||
+		c.bgpPolicyState.localASN != localASN ||
+		c.bgpPolicyState.routerID != routerID
+
+	if needUpdateBGPServer {
+		if c.bgpPolicyState != nil {
+			// Stop the current BGP server.
+			if err := c.bgpPolicyState.bgpServer.Stop(c.ctx); err != nil {
+				return fmt.Errorf("failed to stop current BGP server: %w", err)
+			}
+			// Reset the BGPPolicy state to nil.
+			c.bgpPolicyState = nil
+		}
+
+		// Create a new BGP server.
+		bgpServer := c.newBGPServerFn(&bgp.GlobalConfig{
+			ASN:        uint32(localASN),
+			RouterID:   routerID,
+			ListenPort: listenPort,
+		})
+
+		// Start the new BGP server.
+		if err := bgpServer.Start(c.ctx); err != nil {
+			return fmt.Errorf("failed to start BGP server: %w", err)
+		}
+
+		// Initialize the BGPPolicy state to store the new BGP server, listen port, local ASN, and router ID.
+		c.bgpPolicyState = &bgpPolicyState{
+			bgpServer:   bgpServer,
+			routerID:    routerID,
+			listenPort:  listenPort,
+			localASN:    localASN,
+			routes:      make(sets.Set[bgp.Route]),
+			peerConfigs: make(map[string]bgp.PeerConfig),
+		}
+	}
+
+	// Reconcile BGP peers.
+	if err := c.reconcileBGPPeers(effectiveBP.Spec.BGPPeers); err != nil {
+		return err
+	}
+
+	// Reconcile BGP advertisements.
+	if err := c.reconcileBGPAdvertisements(effectiveBP.Spec.Advertisements); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcileBGPPeers(bgpPeers []v1alpha1.BGPPeer) error {
+	curPeerConfigs := c.getPeerConfigs(bgpPeers)
+	prePeerConfigs := c.bgpPolicyState.peerConfigs
+	prePeerKeys := sets.KeySet(prePeerConfigs)
+	curPeerKeys := sets.KeySet(curPeerConfigs)
+
+	peerToAddKeys := curPeerKeys.Difference(prePeerKeys)
+	peerToUpdateKeys := sets.New[string]()
+	for peerKey := range prePeerKeys.Intersection(curPeerKeys) {
+		prevPeerConfig := prePeerConfigs[peerKey]
+		curPeerConfig := curPeerConfigs[peerKey]
+		if !reflect.DeepEqual(prevPeerConfig, curPeerConfig) {
+			peerToUpdateKeys.Insert(peerKey)
+		}
+	}
+	peerToDeleteKeys := prePeerKeys.Difference(curPeerKeys)
+
+	bgpServer := c.bgpPolicyState.bgpServer
+	for key := range peerToAddKeys {
+		peerConfig := curPeerConfigs[key]
+		if err := bgpServer.AddPeer(c.ctx, peerConfig); err != nil {
+			return err
+		}
+		c.bgpPolicyState.peerConfigs[key] = peerConfig
+	}
+	for key := range peerToUpdateKeys {
+		peerConfig := curPeerConfigs[key]
+		if err := bgpServer.UpdatePeer(c.ctx, peerConfig); err != nil {
+			return err
+		}
+		c.bgpPolicyState.peerConfigs[key] = peerConfig
+	}
+	for key := range peerToDeleteKeys {
+		peerConfig := prePeerConfigs[key]
+		if err := bgpServer.RemovePeer(c.ctx, peerConfig); err != nil {
+			return err
+		}
+		delete(c.bgpPolicyState.peerConfigs, key)
```

Comment:
Instead of using `c.ctx` directly for the add / update / remove peer operations, I think it may make sense to create a child context with a reasonable timeout using https://pkg.go.dev/context#WithTimeout

Reply:
I'm agreed that it will be better to have a reasonable timeout for calling add/update/remove peer operations, but I'm not sure about how to decide timeout.